File: src/main/java/com/rcoverzola/ejers/AgeTaxAction.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.rcoverzola.ejers;

import com.util.ProgramUtil;
import java.time.LocalDate;
import java.time.Period;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;

/**
 *
 * @author Roger Coverzola Bogado
 */
public class AgeTaxAction extends ProgramUtil {

  private LocalDate dateCurrent;
  private HashMap<String, String> objects;
  private static int OLD_AGE = 18;
  private static int MAX_AGE = 150;
  private int CONT = 1;
  private static int LIMIT_TRX = 1000000;
  private int monthCurrent;

  public AgeTaxAction() {
    String[] message = {"Ingrese su fecha de nacimiento (dd/mm/yyyy): ", "Igrese el ingreso para el mes ", "Fecha de naciomiento no valida", "Si, ", "No, ", "debe tributar","Adios!!"};
    init(message);
    this.objects = new HashMap<String, String>();
    this.dateCurrent = LocalDate.now();
    this.monthCurrent = dateCurrent.getMonthValue();
  }

  private void onErrorDate() {
    System.out.println(super.MESSAGES[2]);
    super.EXIT = false;
  }

  private String getNameMoths(int numMohth) {
    switch (numMohth) {
      case 1:
        return "Enero";
      case 2:
        return "Febrero";
      case 3:
        return "Marzo";
      case 4:
        return "Abril";
      case 5:
        return "Mayo";
      case 6:
        return "Junio";
      case 7:
        return "Julio";
      case 8:
        return "Agosto";
      case 9:
        return "Septiembre";
      case 10:
        return "Octubre";
      case 11:
        return "Noviembre";
      case 12:
        return "Diciembre";
      default:
        super.EXIT = false;
        return null;
    }
  }

  private void cal() {
    Double aux = 0.0;
    String res = super.MESSAGES[4];
    for (String value : this.objects.values()) {
      aux += Double.valueOf(value);
    }
    if (aux >= this.LIMIT_TRX) {
      res=super.MESSAGES[3];
    } 
      System.out.println(res+super.MESSAGES[5]);
  }

  public void calTaxAction() {
    System.out.println(super.MESSAGES[0]);
    String birthDateStr = super.scanner.nextLine();

    LocalDate birthDate = LocalDate.parse(birthDateStr, DateTimeFormatter.ofPattern("dd/MM/yyyy"));
    Period dif = Period.between(birthDate, this.dateCurrent);
    if (dif.getYears() < this.OLD_AGE || dif.getYears() >= this.MAX_AGE) {
      onErrorDate();
    }
    while (this.EXIT) {
      System.out.println(super.MESSAGES[1] + getNameMoths(this.CONT) + ":");
      Double value = super.scanner.nextDouble();
      this.objects.put(getNameMoths(this.CONT), String.valueOf(value));
      if (this.CONT == this.monthCurrent) {
        cal();
        super.scanner.close();
        System.out.println(this.objects.toString());
        System.out.println(this.MESSAGES[6]);
        this.EXIT = false;
      } else {
        this.CONT += 1;
      }
    }

  }

}

File: src/AgeTaxAction.ts
import { ProgramUtil } from './ProgramUtil';

export class AgeTaxAction extends ProgramUtil {
    private dateCurrent: Date;
    private objects: Map<string, string>;
    private static readonly OLD_AGE: number = 18;
    private static readonly MAX_AGE: number = 150;
    private CONT: number = 1;
    private static readonly LIMIT_TRX: number = 1000000;
    private monthCurrent: number;

    constructor() {
        super();
        const message: string[] = [
            "Ingrese su fecha de nacimiento (dd/mm/yyyy): ", 
            "Igrese el ingreso para el mes ", 
            "Fecha de naciomiento no valida", 
            "Si, ", 
            "No, ", 
            "debe tributar",
            "Adios!!"
        ];
        this.init(...message);
        this.objects = new Map<string, string>();
        this.dateCurrent = new Date();
        this.monthCurrent = this.dateCurrent.getMonth() + 1; // getMonth() devuelve 0-11
    }

    private onErrorDate(): void {
        console.log(this.MESSAGES[2]);
        this.EXIT = false;
    }

    private getNameMoths(numMohth: number): string | null {
        switch (numMohth) {
            case 1:
                return "Enero";
            case 2:
                return "Febrero";
            case 3:
                return "Marzo";
            case 4:
                return "Abril";
            case 5:
                return "Mayo";
            case 6:
                return "Junio";
            case 7:
                return "Julio";
            case 8:
                return "Agosto";
            case 9:
                return "Septiembre";
            case 10:
                return "Octubre";
            case 11:
                return "Noviembre";
            case 12:
                return "Diciembre";
            default:
                this.EXIT = false;
                return null;
        }
    }

    private cal(): void {
        let aux: number = 0;
        let res: string = this.MESSAGES[4] || '';
        
        for (const value of this.objects.values()) {
            aux += parseFloat(value);
        }
        
        if (aux >= AgeTaxAction.LIMIT_TRX) {
            res = this.MESSAGES[3] || '';
        } 
        
        console.log(res + (this.MESSAGES[5] || ''));
    }

    public calTaxAction(): void {
        console.log(this.MESSAGES[0]);
        
        this.scanner.question('', (birthDateStr: string) => {
            // Parsear fecha en formato dd/mm/yyyy
            const parts = birthDateStr.split('/');
            if (parts.length === 3) {
                const birthDate = new Date(parseInt(parts[2] || '0'), parseInt(parts[1] || '1') - 1, parseInt(parts[0] || '1'));
            
            const today = new Date();
            const diffTime = Math.abs(today.getTime() - birthDate.getTime());
            const diffYears = Math.ceil(diffTime / (1000 * 60 * 60 * 24 * 365.25));
            
            if (diffYears < AgeTaxAction.OLD_AGE || diffYears >= AgeTaxAction.MAX_AGE) {
                this.onErrorDate();
            }
            
            if (this.EXIT) {
                this.askForIncome();
            }
            } else {
                console.log("Formato de fecha inválido. Use dd/mm/yyyy");
                this.onErrorDate();
            }
        });
    }

    private askForIncome(): void {
        if (!this.EXIT) {
            this.scanner.close();
            return;
        }
        
        const monthName = this.getNameMoths(this.CONT);
        if (!monthName) {
            this.scanner.close();
            return;
        }
        
        console.log(this.MESSAGES[1] + monthName + ":");
        
        this.scanner.question('', (value: string) => {
            this.objects.set(monthName, value);
            
            if (this.CONT === this.monthCurrent) {
                this.cal();
                this.scanner.close();
                console.log(Object.fromEntries(this.objects));
                console.log(this.MESSAGES[6]);
                this.EXIT = false;
            } else {
                this.CONT += 1;
                this.askForIncome();
            }
        });
    }
}
